### PR TITLE
P-Chain height progress subscription for subnets

### PR DESCRIPTION
**Pull request by yacovm**
## Why this should be merged

The consumer of this feature is expected to be a Simplex instance, which needs to learn about P-chain height advancement.

The reason is that in avalanche (both in snowman and Simplex) a consensus instance only builds blocks whenever there are pending transactions to be included.

However, unlike snowman - A Simplex L1 also listens in parallel to the P-chain reconfiguring its validator set in order to initiate an epoch change. Therefore, Simplex might need to build a block without transactions when it detects a validator set change in the P-chain (if there are no transactions in the L1's mempool).

To that end, this commit adds a mechanism for a Simplex instance to subscribe for updates about the P-chain advancing, so it can efficiently probe the P-chain's validator set and determine if it needs to initiate an epoch change, while the alternative is to periodically probe the P-chain's height, which is inefficient and incurs a delay.

## How this works

This commit adds a mechanism in which the P-chain can publish to an L1 instance a notification about its P-chain height being updated.

Each L1 can subscribe for notifications starting from a P-chain height (non inclusive).

This commit includes the implementation of the pub/sub mechanism, as well as wiring for publishing the height advancement of the P-chain, but without the consumer (subscriber) which will be added later.

## How this was tested

Added unit tests

### CODE REVIEW

File: utils/lock/progress_test.go [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+
+	select {
+	case <-done:
+	case <-time.After(time.Minute):
+		require.Fail(t, "WaitForProgress should have returned immediately")
+	}
+}
+
+func TestProgressSubscriptionWaiterBlocksUntilProgressAdvances(t *testing.T) {
+	ps := NewProgressSubscription[int](0)
```
Comment (samliok) [inline utils/lock/progress_test.go]:
```suggestion
	ps := NewProgressSubscription(0)
```

File: utils/lock/progress.go [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+package lock
+
+import (
+	"cmp"
+	"context"
+	"sync"
+)
+
+// ProgressSubscription allows a caller to subscribe to progress updates.
+type ProgressSubscription[T cmp.Ordered] struct {
```
Comment (samliok) [inline utils/lock/progress.go:13]: no action needed: This name feels weird to me. It's basically a wrapper over a conditional variable with a custom >= check. I dont have a better name for it, so I think its fine.

File: utils/lock/progress_test.go (hunk 1 above)
Comment (samliok) [reply utils/lock/progress_test.go]: same with other tests

File: chains/manager.go [hunk 3]
```diff
@@ -1243,6 +1248,16 @@ func (m *manager) createSnowmanChain(
...
 	}
 
+	if ctx.ChainID == constants.PlatformChainID {
+		lastAcceptedHeight, err := getLastAcceptedHeight(vm)
+		if err != nil {
+			return nil, fmt.Errorf("error while getting last accepted height: %w", err)
+		}
+
+		m.pChainProgress = lock.NewProgressSubscription[uint64](lastAcceptedHeight)
+		pchainProgressUpdater = m.pChainProgress
```
Comment (samliok) [inline chains/manager.go:1258]: no action needed: would be nice to test this with an integration test or something, but i think that would be too much of a hassle.

File: utils/lock/progress.go [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+}
+
+// WaitForProgress blocks until the progress of this ProgressSubscription is above the given value,
+// or until the given context is cancelled.
+func (ps *ProgressSubscription[T]) WaitForProgress(ctx context.Context, pos T) {
+	ps.lock.Lock()
+	defer ps.lock.Unlock()
+
+	for pos >= ps.progress {
+		if err := ps.signal.Wait(ctx); err != nil {
```
Comment (DracoLi) [inline utils/lock/progress.go:45]: should this return the error to distinguish "progress reached" from "context cancelled."?
While the caller can check `ctx.Err()` after the call, returning error can make the API unambiguous - `nil` means progress was reached, non-nil means it wasn't.

File: utils/lock/progress.go [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+// NewProgressSubscription returns a new ProgressSubscription with the given initial progress.
+func NewProgressSubscription[T cmp.Ordered](initialProgress T) *ProgressSubscription[T] {
+	var ps ProgressSubscription[T]
+	ps.signal = NewCond(&ps.lock)
+	ps.progress = initialProgress
+	return &ps
+}
+
+// SetProgress updates the progress of this ProgressSubscription to the given value.
+// This will unblock any calls to WaitForProgress that are waiting for a progress value above the given value.
```
Comment (DracoLi) [inline utils/lock/progress.go:28]: should this enforce progress can never go down? nothing preventing SetProgress(5), then SetProgress(3).

File: chains/manager.go [hunk 6]
```diff
@@ -1243,6 +1248,16 @@ func (m *manager) createSnowmanChain(
 		return nil, err
 	}
 
+	if ctx.ChainID == constants.PlatformChainID {
+		lastAcceptedHeight, err := getLastAcceptedHeight(vm)
+		if err != nil {
+			return nil, fmt.Errorf("error while getting last accepted height: %w", err)
+		}
+
+		m.pChainProgress = lock.NewProgressSubscription[uint64](lastAcceptedHeight)
```
Comment (DracoLi) [inline chains/manager.go]: would the `uint64` be inferred?
```suggestion
		m.pChainProgress = lock.NewProgressSubscription(lastAcceptedHeight)
```

File: utils/lock/progress_test.go (hunk 1 above)
Comment (yacovm) [reply utils/lock/progress_test.go]: https://github.com/ava-labs/avalanchego/pull/5110/changes/51240cf2e4620243f28bd4a35eb60e9e7521e7f6

File: chains/manager.go (hunk 3 above)
Comment (yacovm) [reply chains/manager.go:1258]: We don't have tests for the manager and even Claude Code flat out refuses to make unit tests for this without heavy refactoring:

> The reality is that createSnowmanChain is ~300 lines with dozens of dependencies — it's not practically unit-testable without massive setup.

File: snow/engine/snowman/engine_test.go [hunk 7]
```diff
@@ -3301,3 +3301,62 @@ func TestEngineAcceptedHeight(t *testing.T) {
 	require.Equal(blk1.Height(), h1)
 	require.Equal(blk2.Height(), h2)
 }
+
+func TestPChainProgressUpdaterCalledOnAccept(t *testing.T) {
+	require := require.New(t)
```
Comment (DracoLi) [inline snow/engine/snowman/engine_test.go:3306]: nit: I believe we are moving towards just using standalone require.Equal instead of `require.New`

Comment (yacovm) [reply snow/engine/snowman/engine_test.go:3306]: That's the style of the tests in the package. I think that if we move to something else we should do it holistically and atomically, not periodically.

File: utils/lock/progress.go (hunk 5 above)
Comment (yacovm) [reply utils/lock/progress.go:28]: That's also what CC told me. I don't think we should guard against malicious input. If progress can go down, we are in bigger trouble than us getting stuck 😉

File: utils/lock/progress.go (hunk 4 above)
Comment (yacovm) [reply utils/lock/progress.go:45]: I admit I am a bit on the fence about this. 

On the one hand, indeed if we return we don't know if the context was cancelled or not. On the other hand, currently the only [consumer](https://github.com/ava-labs/Simplex/blob/reconfig/msm/build_decision.go#L71-L87) of this API is Simplex, and it doesn't care why we returned:

```
func (bbd *blockBuildingDecider) waitForPChainChangeOrPendingBlock(ctx context.Context, pChainHeight uint64) {
	pChainAwareContext, cancel := context.WithCancel(ctx)

	var wg sync.WaitGroup
	wg.Add(1)

	defer wg.Wait()
	defer cancel()

	go func() {
		defer wg.Done()
		bbd.pChainlistener.WaitForProgress(pChainAwareContext, pChainHeight)
		cancel()
	}()

	bbd.waitForPendingBlock(pChainAwareContext)
}
```

All we need is to make sure that if we have made progress in the P-chain then we get unblocked. 

I guess I can add an error return value and log it. I'll add a TODO and will address this when I add logging, so in the meantime I'll add the return value to this PR. 👍

File: chains/manager.go (hunk 6 above)
Comment (yacovm) [reply chains/manager.go]: yes

File: utils/lock/progress.go (hunk 5 above)
Comment (DracoLi) [reply utils/lock/progress.go:28]: maybe worth a note here then to call this out?

Comment (yacovm) [reply utils/lock/progress.go:28]: https://github.com/ava-labs/avalanchego/pull/5110/commits/7d3f3cf85952fb4395e74c7586f394a0a0522b5b

File: snow/engine/snowman/engine_test.go [hunk 8]
```diff
@@ -3301,3 +3301,62 @@ func TestEngineAcceptedHeight(t *testing.T) {
...
+
+	// Vote for the block to accept it
+	require.NoError(te.Chits(t.Context(), vdr, *queryRequestID, blk.ID(), blk.ID(), blk.ID(), blk.Height()))
+
+	require.Equal(snowtest.Accepted, blk.Status)
+	require.True(progressUpdated)
+	require.Equal(blk.Height(), updatedHeight)
+}
+
+type mockPChainProgressUpdater struct {
```
Comment (joshua-kim) [inline snow/engine/snowman/engine_test.go:3356]: Why did we end up using mocks here? This is a testable type.

File: utils/lock/progress.go [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+	ps.lock.Lock()
+	defer ps.lock.Unlock()
+
+	ps.progress = progress
+	ps.signal.Broadcast()
+}
+
+// WaitForProgress blocks until the progress of this ProgressSubscription is above the given value,
+// or until the given context is cancelled.
+func (ps *ProgressSubscription[T]) WaitForProgress(ctx context.Context, pos T) error {
```
Comment (joshua-kim) [inline utils/lock/progress.go:40]: This is basically a context-aware semaphore/channel. Why isn't a channel sufficient here/do we need this type given that only simplex is using it? One nice-ity of channels is that we can use `select` to write simple concurrent code. I generally try to be intentional about creating public apis/make them private where possible.

File: utils/lock/progress.go [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+// ProgressSubscription allows a caller to subscribe to progress updates.
+type ProgressSubscription[T cmp.Ordered] struct {
+	signal   *Cond
+	lock     sync.Mutex
+	progress T
+}
+
+// NewProgressSubscription returns a new ProgressSubscription with the given initial progress.
+func NewProgressSubscription[T cmp.Ordered](initialProgress T) *ProgressSubscription[T] {
```
Comment (joshua-kim) [inline utils/lock/progress.go:20]: The generic feels premature given that we are only using it to advertise height update events.

File: utils/lock/progress.go (hunk 2 above)
Comment (joshua-kim) [reply utils/lock/progress.go:13]: +1, the package is also the abstraction so `lock.ProgressSubscription` is unclear to me.

File: utils/lock/progress_test.go [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+	go func() {
+		require.NoError(t, ps.WaitForProgress(t.Context(), 5))
+		close(done)
+	}()
+
+	select {
+	case <-done:
+	case <-time.After(time.Minute):
+		require.Fail(t, "WaitForProgress should have returned immediately")
+	}
```
Comment (joshua-kim) [inline utils/lock/progress_test.go:29]: We don't need to write these timeout cases -- the test runtime has a default timeout.

File: utils/lock/progress_test.go [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+		err := ps.WaitForProgress(ctx, 5)
+		require.ErrorIs(t, err, context.Canceled)
+		close(done)
+	}()
+
+	// The goroutine should be blocked.
+	select {
+	case <-done:
+		require.Fail(t, "WaitForProgress should not have returned yet")
+	case <-time.After(50 * time.Millisecond):
```
Comment (joshua-kim) [inline utils/lock/progress_test.go:83]: This is flakey/racey -- we're not guaranteed that the goroutine is unblocked by the runtime.

File: utils/lock/progress.go (hunk 9 above)
Comment (yacovm) [reply utils/lock/progress.go:40]: How can we notify many consumers with a channel? We can have many instances of Simplex.

File: chains/manager.go (hunk 3 above)
Comment (joshua-kim) [reply chains/manager.go:1258]: It's not possible to test this because there is no usage of this yet -- I think it's difficult to review this without knowledge of how `WaitForProgress` is either used or being a part of the diff

File: utils/lock/progress.go (hunk 10 above)
Comment (yacovm) [reply utils/lock/progress.go:20]: Well since I put it in utils, I figured I might make it generic

File: chains/manager.go [hunk 13]
```diff
@@ -1598,3 +1614,17 @@ func (m *manager) getOrMakeVMGatherer(vmID ids.ID) (metrics.MultiGatherer, error
 	m.vmGatherer[vmID] = vmGatherer
 	return vmGatherer, nil
 }
+
+func getLastAcceptedHeight(vm block.ChainVM) (uint64, error) {
+	lastAcceptedBlock, err := vm.LastAccepted(context.Background())
```
Comment (joshua-kim) [inline chains/manager.go:1619]: This breaks contexts (e.g tracing, cancelation) -- we should be always be accepting the parent context or using `TODO`. Typically `Background` is used in `main` and children are created off of it throughout the lifetime of an application

File: utils/lock/progress_test.go (hunk 11 above)
Comment (yacovm) [reply utils/lock/progress_test.go:29]: Do you suggest to not have a select at all and just let the test timeout? I think it's easier to understand a regression if we have more context on the failure.

File: utils/lock/progress.go (hunk 9 above)
Comment (joshua-kim) [reply utils/lock/progress.go:40]: My thought was that we could have implemented some type of fan-out pattern with channels. The way this is implemented this will unblock all consumers even if they haven't finished handling a notification (I'm not sure what that means without knowing how the consumer works though).

File: utils/lock/progress_test.go (hunk 12 above)
Comment (yacovm) [reply utils/lock/progress_test.go:83]: We are checking that we are *not* unblocked. We should be blocked, not unblocked.

Comment (joshua-kim) [reply utils/lock/progress_test.go:83]: My callout is that we are not guaranteed anything -- it's possible that this goroutine hasn't even started running yet. If we want to test cancelation behavior one way of writing this without introducing time-based sleeps would be something like this:

```
func TestProgressSubscriptionContextCancellationUnblocksWaiter(t *testing.T) {
	ps := NewProgressSubscription(0)

	ctx, cancel := context.WithCancel(t.Context())

	eg := &errgroup.Group{}
	eg.Go(func() error {
		return ps.WaitForProgress(ctx, 5)
	})

	cancel()

	require.ErrorIs(t, eg.Wait(), context.Canceled)
}
```

File: utils/lock/progress.go (hunk 5 above)
Comment (joshua-kim) [reply utils/lock/progress.go:28]: We could add a defensive check here to limit the blast radius of failure -- it is trivial to add a panic/error and it would be better to fail here than to silently fail downstream in consensus.

File: utils/lock/progress.go (hunk 9 above)
Comment (yacovm) [reply utils/lock/progress.go:40]: So, in the consumer code I have a context, so I need a context here too. 

> The way this is implemented this will unblock all consumers even if they haven't finished handling a notification (I'm not sure what that means without knowing how the consumer works though).

Yeah indeed it will unblock all consumers, however we have only a single P-chain, so if a consumer is waiting then we need to unblock it whenever the P-chain changes. This works as desired.

File: snow/engine/snowman/engine_test.go (hunk 8 above)
Comment (yacovm) [reply snow/engine/snowman/engine_test.go:3356]: > This is a testable type.

What do you mean by "this is a testable type" ? 

Would you prefer I'd use the real `ProgressSubscription` ?

File: utils/lock/progress_test.go (hunk 12 above)
Comment (yacovm) [reply utils/lock/progress_test.go:83]: oh I see, I didn't know the `errgroup` trick. 

Updated to adopt your solution, thanks!

File: chains/manager.go (hunk 13 above)
Comment (yacovm) [reply chains/manager.go:1619]: Fixed

File: snow/engine/snowman/config.go [hunk 14]
```diff
@@ -17,12 +17,13 @@ import (
...
-	PartialSync         bool
+	Ctx                   *snow.ConsensusContext
+	VM                    block.ChainVM
+	Sender                common.Sender
+	Validators            validators.Manager
+	ConnectedValidators   tracker.Peers
+	Params                snowball.Parameters
+	Consensus             snowman.Consensus
+	PartialSync           bool
+	PChainProgressUpdater PChainProgressUpdater
```
Comment (StephenButtolph) [inline snow/engine/snowman/config.go:28]: I don't think this makes sense. `PChainProgressUpdater` living in `snowman.Config` is a backwards dependency.

There are many snowman instances, but only one P-chain. So why should every snowman instance be able to take in a P-chain specific progress updater?

I think any such progress manager should either live in the P-chain directly, or be more agnostic to the specific chain it is running.

File: snow/engine/snowman/engine.go [hunk 15]
```diff
@@ -1100,10 +1100,19 @@ func (e *Engine) addUnverifiedBlockToConsensus(
 		zap.Stringer("blkID", blkID),
 		zap.Uint64("height", blkHeight),
 	)
+
+	onAccept := func() {}
+	if e.PChainProgressUpdater != nil {
+		onAccept = func() {
+			e.PChainProgressUpdater.SetProgress(blkHeight)
+		}
+	}
```
Comment (StephenButtolph) [inline snow/engine/snowman/engine.go:1109]: Why do we introduce this indirection here? By passing in a `func()` to `memoryBlock`, we lose the ability to look at `memoryBlock.Accept` and understand what the code is doing.

I'm not saying this pattern never makes sense... But we aren't dealing with a circular dependency here... We should just provide the progress tracker directly to the `memoryBlock` so that we know we are updating any listeners with the last accepted height.

File: chains/manager.go [hunk 16]
```diff
@@ -1114,6 +1117,8 @@ func (m *manager) createSnowmanChain(
 		messageSender = sender.Trace(messageSender, m.Tracer)
 	}
 
+	var pchainProgressUpdater smeng.PChainProgressUpdater
+
```
Comment (StephenButtolph) [inline chains/manager.go:1121]: Why is this defined here rather than where it is initialized?

File: chains/manager.go [hunk 17]
```diff
@@ -1243,6 +1248,16 @@ func (m *manager) createSnowmanChain(
 		return nil, err
 	}
 
+	if ctx.ChainID == constants.PlatformChainID {
+		lastAcceptedHeight, err := getLastAcceptedHeight(vm)
+		if err != nil {
+			return nil, fmt.Errorf("error while getting last accepted height: %w", err)
+		}
+
+		m.pChainProgress = lock.NewProgressSubscription(lastAcceptedHeight)
```
Comment (StephenButtolph) [inline chains/manager.go:1257]: If we aren't tying this to the VM, why are we initializing it in `createSnowmanChain` and not `New` like the other `manager` fields?

File: snow/engine/snowman/engine.go [hunk 18]
```diff
@@ -1200,3 +1209,9 @@ func (e *Engine) isDecided(blk snowman.Block) bool {
 	parentID := blk.Parent()
 	return parentHeight == lastAcceptedHeight && parentID != lastAcceptedID // the parent was rejected
 }
+
+// PChainProgressUpdater is used to update the P-Chain height progress of this chain.
+// It's only used by the engine instance that runs the P-chain.
+type PChainProgressUpdater interface {
+	SetProgress(height uint64)
+}
```
Comment (StephenButtolph) [inline snow/engine/snowman/engine.go:1217]: Why does this interface exist?

File: utils/lock/progress_test.go [hunk 19]
```diff
@@ -0,0 +1,157 @@
...
+	"github.com/stretchr/testify/require"
+)
+
+func TestProgressSubscriptionInitialProgressUnblocksWaiterImmediately(t *testing.T) {
+	ps := NewProgressSubscription[int](10)
+
+	// Waiting for a value below the initial progress should return immediately.
+	done := make(chan struct{})
+	go func() {
+		require.NoError(t, ps.WaitForProgress(t.Context(), 5))
```
Comment (StephenButtolph) [inline utils/lock/progress_test.go:21]: `require` calls `FailNow` which panics. `require` should not be used outside of the main testing goroutine.

File: utils/lock/progress_test.go [hunk 20]
```diff
@@ -0,0 +1,157 @@
...
+		close(done)
+	}()
+
+	select {
+	case <-done:
+	case <-time.After(time.Minute):
+		require.Fail(t, "WaitForProgress should have returned immediately")
+	}
+}
```
Comment (StephenButtolph) [inline utils/lock/progress_test.go:30]: Why do we do all of this logic?
```suggestion
func TestProgressSubscriptionInitialProgressUnblocksWaiterImmediately(t *testing.T) {
	ps := NewProgressSubscription[int](10)
	require.NoError(t, ps.WaitForProgress(t.Context(), 5))
}
```

File: utils/lock/progress_test.go [hunk 21]
```diff
@@ -0,0 +1,157 @@
...
+	ps.SetProgress(10)
+
+	select {
+	case <-done:
+	case <-time.After(time.Minute):
+		require.Fail(t, "WaitForProgress should have returned after SetProgress")
+	}
+
+	wg.Wait()
+}
```
Comment (StephenButtolph) [inline utils/lock/progress_test.go:61]: Why does this wg exist? Also, the inner delay really doesn't add anything imo.
```suggestion
func TestProgressSubscriptionWaiterBlocksUntilProgressAdvances(t *testing.T) {
	ps := NewProgressSubscription(0)

	go ps.SetProgress(10)
	require.NoError(t, ps.WaitForProgress(t.Context(), 5))
}
```

File: utils/lock/progress_test.go [hunk 22]
```diff
@@ -0,0 +1,157 @@
...
+
+	cancel()
+
+	select {
+	case <-done:
+	case <-time.After(time.Minute):
+		require.Fail(t, "WaitForProgress should have returned after context cancellation")
+	}
+
+	wg.Wait()
```
Comment (StephenButtolph) [inline utils/lock/progress_test.go:94]: Similarly, I don't think the inner wait does anything useful.
```suggestion
	ps := NewProgressSubscription(0)

	ctx, cancel := context.WithCancel(t.Context())
	go cancel()
	require.ErrorIs(t, ps.WaitForProgress(ctx, 5), context.Canceled)
```

File: utils/lock/progress_test.go [hunk 23]
```diff
@@ -0,0 +1,157 @@
...
+	case <-time.After(50 * time.Millisecond):
+	}
+
+	ps.SetProgress(6)
+
+	select {
+	case <-done:
+	case <-time.After(time.Minute):
+		require.Fail(t, "WaitForProgress should have returned after progress exceeded pos")
+	}
```
Comment (StephenButtolph) [inline utils/lock/progress_test.go:156]: We have already tested that `SetProgress` can unblock, we are just checking the boundary case here.
```suggestion
	ps := NewProgressSubscription(5)

	ctx, cancel := context.WithCancel(t.Context())
	go cancel()
	require.ErrorIs(t, ps.WaitForProgress(ctx, 5), context.Canceled)
```

File: utils/lock/progress_test.go [hunk 24]
```diff
@@ -0,0 +1,157 @@
...
+	for i := range numWaiters {
+		select {
+		case <-done[i]:
+		case <-time.After(time.Minute):
+			require.Fail(t, "waiter should have been unblocked")
+		}
+	}
+
+	wg.Wait()
+}
```
Comment (StephenButtolph) [inline utils/lock/progress_test.go:132]: This seems like just a more robust version of `TestProgressSubscriptionWaiterBlocksUntilProgressAdvances`. Can we just delete `TestProgressSubscriptionWaiterBlocksUntilProgressAdvances` entirely and use this function?

Comment (StephenButtolph) [inline utils/lock/progress_test.go:132]: We should be using the more robust `synctest` package for stuff like this:
```golang
// Give goroutines time to start waiting.
time.Sleep(50 * time.Millisecond)
```

Also, we shouldn't mix synchronization primitives.

```suggestion
func TestProgressSubscriptionMultipleWaitersUnblockedBySingleProgress(t *testing.T) {
	synctest.Test(t, func(t *testing.T) {
		ps := NewProgressSubscription(0)

		var eg errgroup.Group
		for range 5 {
			eg.Go(func() error {
				return ps.WaitForProgress(t.Context(), 0)
			})
		}
		synctest.Wait() // Ensure all the waiters are waiting.

		ps.SetProgress(1)
		require.NoError(t, eg.Wait())
	})
}
```

File: utils/lock/progress_test.go
Comment (StephenButtolph) [inline utils/lock/progress_test.go:1]: I think there are a lot of duplicate / similar tests here. I really feel like just:
```golang
func TestProgressSubscription_NoUpdate(t *testing.T) {
	tests := []struct {
		name    string
		initial int
		last    int
		want    error
	}{
		{
			name:    "initial > last",
			initial: 1,
			last:    0,
			want:    nil,
		},
		{
			name:    "initial == last",
			initial: 0,
			last:    0,
			want:    context.Canceled,
		},
		{
			name:    "initial < last",
			initial: 0,
			last:    1,
			want:    context.Canceled,
		},
	}
	for _, test := range tests {
		t.Run(test.name, func(t *testing.T) {
			ps := NewProgressSubscription(test.initial)
			ctx, cancel := context.WithCancel(t.Context())
			go cancel()
			require.ErrorIs(t, ps.WaitForProgress(ctx, test.last), test.want)
		})
	}
}
```

And `TestProgressSubscriptionMultipleWaitersUnblockedBySingleSetProgress` provide the same level of coverage with significantly less duplication.

File: snow/engine/snowman/config.go (hunk 14 above)
Comment (yacovm) [reply snow/engine/snowman/config.go:28]: > There are many snowman instances, but only one P-chain. So why should every snowman instance be able to take in a P-chain specific progress updater?

My plan was to expand the interface to include the subscriber part, once we merge that part of Simplex.

Then, every chain would have a progress updater, and only the P-chain would `SetProgress(height uint64)` and the others that run Simplex would pass on `WaitForProgress(ctx context.Context, pos T) error` or the interface itself (I don't remember which one it is)

File: snow/engine/snowman/engine.go (hunk 15 above)
Comment (yacovm) [reply snow/engine/snowman/engine.go:1109]: Because only the P-chain needs to update the progress upon accept. 

If I provide the progress tracker directly into the memory block, I will need to check if this is a memory block of the P-chain or not, and this is further complication.

File: snow/engine/snowman/config.go (hunk 14 above)
Comment (yacovm) [reply snow/engine/snowman/config.go:28]: Removed it from the snowman config and added it to the consensus context. It's not agnostic to the chain it is running in.

File: snow/engine/snowman/engine.go (hunk 15 above)
Comment (yacovm) [reply snow/engine/snowman/engine.go:1109]: The memory block now always [updates](https://github.com/ava-labs/avalanchego/pull/5152/changes/b376c6eebec58820e8fd9cfcd4866083cb7f34eb#diff-b367891d6b881489ef8250edec01fc2827e93664ac21fb5ac8314c3007753dcaL31) the chain height regardless of the chain.

File: chains/manager.go (hunk 16 above)
Comment (yacovm) [reply chains/manager.go:1121]: [Removed](https://github.com/ava-labs/avalanchego/pull/5152/changes/b376c6eebec58820e8fd9cfcd4866083cb7f34eb#diff-e0a11726ad4d90955301188800ed0fdb8573df2095e80e99b5929614eca9c592L1120) this.

File: chains/manager.go (hunk 17 above)
Comment (yacovm) [reply chains/manager.go:1257]: [Removed](https://github.com/ava-labs/avalanchego/pull/5152/changes/b376c6eebec58820e8fd9cfcd4866083cb7f34eb#diff-e0a11726ad4d90955301188800ed0fdb8573df2095e80e99b5929614eca9c592L256) the field from the manager.

File: snow/engine/snowman/engine.go (hunk 18 above)
Comment (yacovm) [reply snow/engine/snowman/engine.go:1217]: I thought it was a cleaner abstraction given that we don't use the other method just yet. Was planning on removing it once we integrate Simplex.

[Removed](https://github.com/ava-labs/avalanchego/pull/5152/changes/b376c6eebec58820e8fd9cfcd4866083cb7f34eb#diff-1299463859a867cd625ef56c21a2f2fb4d30fc9b43b9e13c04350d0b76472f32L1215-L1219) it.

File: utils/lock/progress_test.go (hunk 19 above)
Comment (yacovm) [reply utils/lock/progress_test.go:21]: Removed the test as asked [here](https://github.com/ava-labs/avalanchego/pull/5110#discussion_r3018404328).

File: utils/lock/progress_test.go (hunk 20 above)
Comment (yacovm) [reply utils/lock/progress_test.go:30]: Removed the test as asked [here](https://github.com/ava-labs/avalanchego/pull/5110#discussion_r3018404328).

File: utils/lock/progress_test.go (hunk 21 above)
Comment (yacovm) [reply utils/lock/progress_test.go:61]: I wanted to test that WaitForProgress doesn't immediately return when the height hasn't been reached yet. 

Removed the test as asked [here](https://github.com/ava-labs/avalanchego/pull/5110#discussion_r3018404328).

File: utils/lock/progress_test.go (hunk 22 above)
Comment (yacovm) [reply utils/lock/progress_test.go:94]: Removed the test as asked [here](https://github.com/ava-labs/avalanchego/pull/5110#discussion_r3018404328).

File: utils/lock/progress_test.go (hunk 23 above)
Comment (yacovm) [reply utils/lock/progress_test.go:156]: Yes, we are indeed checking the boundary case here. 

Removed the test as asked [here](https://github.com/ava-labs/avalanchego/pull/5110#discussion_r3018404328).

File: utils/lock/progress_test.go (hunk 24 above)
Comment (yacovm) [reply utils/lock/progress_test.go:132]: rewrote the test as [suggested](https://github.com/ava-labs/avalanchego/pull/5110#discussion_r3018377252).

Comment (yacovm) [reply utils/lock/progress_test.go:132]: [rewrote](https://github.com/ava-labs/avalanchego/pull/5152/changes/b376c6eebec58820e8fd9cfcd4866083cb7f34eb#diff-0c8adef8fdacf26dffe0c5a8c9a69f20a9af79d951c6b30c295146363ce4b01eR16-R37) the test as [suggested](https://github.com/ava-labs/avalanchego/pull/5110#discussion_r3018377252).

File: utils/lock/progress_test.go
Comment (yacovm) [reply utils/lock/progress_test.go:1]: [rewrote](https://github.com/ava-labs/avalanchego/pull/5152/changes/b376c6eebec58820e8fd9cfcd4866083cb7f34eb#diff-0c8adef8fdacf26dffe0c5a8c9a69f20a9af79d951c6b30c295146363ce4b01eR40-R73) as suggested
